modules/tribes/tests/client/components/TribesPage.component.test.js

```diff
@@ -0,0 +1,177 @@
+
+
+describe('TribesPage', () => {
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('not signed in', () => {
+
+    it('fetch tribes from api and show tribes on page', async () => {
+      // first no api should be called
+      expect(api.read).toHaveBeenCalledTimes(0);
+
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const tribes = getAllByRole('listitem');
+      expect(tribes).toHaveLength(dummyTribes.length + 1); // there is additionally the Suggest Tribe
+
+      tribes.slice(0, -1).forEach((tribe, i) => {
+        expect(tribe).toHaveTextContent(dummyTribes[i].label);
+      });
+
+      // during the test the api should be called only once
+      expect(api.read).toHaveBeenCalledTimes(1);
+    });
+
+    it('the join button should be a link to tribe page', async () => {
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // join button
+      const tribes = getAllByRole('listitem').slice(0, -1);
+
+      tribes.forEach((tribe, i) => {
+        const button = rtl.getByText(tribe, 'Join');
+        expect(button).toHaveAttribute('href', `/signup?tribe=${dummyTribes[i].slug}`);
+      });
+    });
+
+  });
+
+  describe('signed in', () => {
+
+    it('show tribes on page', async () => {
+      // first no api should be called
+      expect(api.read).toHaveBeenCalledTimes(0);
+
+      const { getAllByRole } = render(<TribesPage user={dummyUser} onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const tribes = getAllByRole('listitem');
+      expect(tribes).toHaveLength(dummyTribes.length + 1); // there is additionally the Suggest Tribe
+
+      tribes.slice(0, -1).forEach((tribe, i) => {
+        expect(tribe).toHaveTextContent(dummyTribes[i].label);
+      });
+
+      // during the test the api should be called only once
+      expect(api.read).toHaveBeenCalledTimes(1);
+    });
+
+    it('user is member of some tribes and not member of others', async () => {
+      const { getAllByText } = render(<TribesPage user={dummyUser} onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const join = getAllByText('Join');
+      const leave = getAllByText('Joined');
+      expect(join).toHaveLength(1);
+      expect(leave).toHaveLength(2);
+    });
+
+    it('click join and send api request', async () => {
+      expect(api.join).toHaveBeenCalledTimes(0);
+      const { getAllByText } = render(<TribesPage user={dummyUser} onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // click the join button
+      const join = getAllByText('Join');
+      expect(join).toHaveLength(1);
+      fireEvent.click(join[0]);
```

Comment:
I _think_ if you use the non-_all_ version (`getByText`) it will automatically thrown an exception if there are more than 1, so can simplify this to:

```js
fireEvent.click(getByText('Join'));
```

... _or_, if combining with `waitForElement` something like:

```js
const join = await waitForElement(() => getByText('Join'));

fireEvent.click(join);
```

... which to me gets closer to what the user is doing ... (I can imagine the `await wait()` could be flakey, for example if another API call had to be added after the first one, it might not wait long enough).

Maybe [`findBy`](https://testing-library.com/docs/dom-testing-library/api-queries#findby) is even better though:

```js
const join = await findByText('Join');

fireEvent.click(join);

// .. or even inline

fireEvent.click(await findByText('Join'));
```

Reply:
Rewritten the test fundamentally. So I'll use what i've learned here later. Thank you!